parser/parser.y

```diff
@@ -4591,14 +4605,53 @@ SelectLockOpt:
 
 // See https://dev.mysql.com/doc/refman/5.7/en/union.html
 UnionStmt:
-	UnionClauseList "UNION" UnionOpt SelectStmt
+	UnionClauseList "UNION" UnionOpt SelectStmtPart1 OrderByOptional SelectStmtLimit SelectLockOpt
 	{
+		st := $4.(*ast.SelectStmt)
 		union := $1.(*ast.UnionStmt)
 		union.Distinct = union.Distinct || $3.(bool)
 		lastSelect := union.SelectList.Selects[len(union.SelectList.Selects)-1]
-		endOffset := parser.endOffset(&yyS[yypt-2])
+		endOffset := parser.endOffset(&yyS[yypt-5])
 		parser.setLastSelectFieldText(lastSelect, endOffset)
-		union.SelectList.Selects = append(union.SelectList.Selects, $4.(*ast.SelectStmt))
+		union.SelectList.Selects = append(union.SelectList.Selects, st)
+		if $5 != nil {
+		    union.OrderBy = $5.(*ast.OrderByClause)
+		}
+		if $6 != nil {
+		    union.Limit = $6.(*ast.Limit)
+		}
+		$$ = union
```

Comment:
$7 is not used?

Reply:
yes. Since the `order by`clause and `limit` clause belong to  the `union` statement, the `SelectLockOpt` belongs to the `union` statement, which means nothing according to `SelectLockOpt` definition in https://dev.mysql.com/doc/refman/5.7/en/innodb-locking-reads.html

Reply:
I think we need to add SelectLockOpt on both side of the union in this case.
What do you think? @coocood

Reply:
@tiancaiamao  what if  the left `SelectStmt` have already set  `SelectLockOpt` ?

Reply:
begin txn1
select a from t where union (xxx whatever) for update

begin txn2
update a xxx
commit txn2

commit txn1  (If for update works on union, this operation should fail in TiDB)

Reply:
We should promote to lock mode  either `SelectLock` exists on left SelectStmt or UnionStmt.

The select lock have something to do with transaction isolation level.
TiDB's Snapshot Isolation read operation will not block write operation, but if `for update` is used, read operation is more strict and you can thought it use a lock to prevent write operation.
@spongedu

Reply:
> We should promote to lock mode either SelectLock exists on left SelectStmt or UnionStmt.
>
@tiancaiamao  what this mean ? Fail to get your point..... @.@


BTW, I test these cases in MySQL 5.7.
In MySQL:
First, execute this in the first tx: 
```sql
mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t  union select * from t2 for update;
+----+----+---------+
| id | tn | type    |
+----+----+---------+
|  1 |  5 | cat     |
|  2 |  6 | dog     |
|  3 |  7 | duck    |
|  4 | 29 | chicken |
|  4 | 99 | chicken |
+----+----+---------+
5 rows in set (0.01 sec)
```

Then, execute this in the second tx:
```sql
mysql> update t  set tn = 31 where id = 4;                                                                                                     Query OK, 1 row affected (0.00 sec)
Rows matched: 1  Changed: 1  Warnings: 0

mysql> update t2  set tn = 31 where id = 4;
ERROR 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
```
the `FOR UPDATE` statement here affect only the second `SelectStmt`.

Another condition:
First, execute this in the first tx: 
```sql
mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t  union select * from t2 limit 100  for update;                                                                          +----+-----+---------+
| id | tn  | type    |
+----+-----+---------+
|  1 |   5 | cat     |
|  2 |   6 | dog     |
|  3 |   7 | duck    |
|  4 | 187 | chicken |
+----+-----+---------+
4 rows in set (0.01 sec)
```

Then, execute this in the second tx:
```sql
mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> update t2 set tn = 1 where id = 1;                                                                                                      Query OK, 1 row affected (0.00 sec)
Rows matched: 1  Changed: 1  Warnings: 0

mysql> update t set tn = 1 where id = 1;
Query OK, 1 row affected (0.00 sec)
Rows matched: 1  Changed: 1  Warnings: 0
```
Both Update works. The `FOR UPDATE` statement here seems work on neither `SelectStmt`

Reply:
Okay, there is some subtle difference about the transaction isolation between TiDB and MySQL.
It's hard to explain here. I'll update the implementation myself after this PR is merged.